### DISCUSSION TITLE
JavaScript To Class Fields or not to Class Fields?

### DISCUSSION THREAD

**Discussion by Ordiel** [lit/lit#3069] [General]
Hi, I am going through the documentation, more specially [Avoiding issues with Class Fields](https://lit.dev/docs/components/properties/#avoiding-issues-with-class-fields) where it states:

> In **JavaScript** you **must not use class fields** when declaring reactive properties. Instead, properties must be initialized in the element constructor

and proceeds to provide the following example:

```javascript
constructor() {
  super();
  this.data = {};
}
```

Which is a bit odd given the very first sentence of that page is (emphasis added):

> Lit components receive input and store their state as _JavaScript class fields or properties._

Specially when that entire page is filled with examples like the one [here](https://lit.dev/docs/components/properties/#public-properties-and-internal-state)

```javascript
static properties = {
  _counter: {state: true};
};

constructor()
  super();
  this._counter = 0;
}
```

Also [here](https://lit.dev/docs/components/properties/#observed-attributes)

```javascript
// observed attribute name is "myvalue"
static properties = {
  myValue: { type: Number },
};

constructor() {
  super();
  this.myValue = 0;
}
```

Worth to mention I am not too savvy on JavaScript yet I did make sure I was looking at the JS examples. Is there something I am missing? or _as long_ as the attributes/fields are wrapped on the `properties` [Class Field](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Public_class_fields) then they are valid?

This all is oriented for me to be able to determine if a class as follows would be good enough:

```javascript
class Person extends LitElement {
  constructor() {
    this.id = 0;
    this.name = null;
    this.lastName = null;
    this.age = 0;
  }
  // Some more stuff here
}
```
Or is it that it needs:

* To have a `static properties` field to describe the class attributes (instance fields)
* Each of those attributes to be objects with the proper [property options](https://lit.dev/docs/components/properties/#property-options) instead of holding raw values

Any guidance or link to documentation clarifying this would be greatly appreciated

**Comment by ynxtynxt** [comment]
As I think I understand. You can use class fields as long as they are **not** reactive properties. Being reactive properties mean that the parent class will generate getters/setters for them. So, a property _x_ will have a pair of getter/setter generated and the element will store the real value of _x_ in ___x_. If you declare it as a class field, the getter/setter won't work.

```js

class X {
    xxx = 5; // This happens when you new() so it will hide the prototype's getter/setter settings.
    
    __xxx = 10;
    set xxx( nxxx ){ // Generated once, into the prototype. 
        this.__xxx = nxxx;
        console.log('never be called');
    }
    get xxx(){
        console.log('never be called');
        return this.__xxx;
    }
}

new X().xxx // 5, not 10
```

Note: With Javascript, you can use `@property()` decorator for class field which is what I do. See [Compiling with Babel](https://lit.dev/docs/tools/publishing/#compiling-with-babel)

**Reply by e111077** [reply]
just a heads up. This guidance might change once decorators are native in JS (currently stage 3)

**Comment by sorvell** [comment]
1. To make a property reactive so that setting it triggers an update/render, you need to either return it in the object set via `static properties` or use the `@property` decorator (when using Typescript or Babel).
2. Native class fields are problematic when used with reactive properties because they set the property value on the element instance, shadowing the reactive property which is defined on the element prototype.
3. When we say class fields, we specifically mean making a property assignment in the class declaration, not in the constructor (e.g. this is a class field, `class Foo { bar = 5; }` and this is not `class Foo { constructor() { this.bar = 5; }}`.
4. Also, be careful setting native [web platform properties](https://www.w3schools.com/tags/ref_standardattributes.asp) like `id` in the element constructor. This value reflects to an attribute and setting it in the constructor will generate an error.

Hope that helps.

**Reply by Ordiel** [reply]
Thanks for the pointers. Specially #4, which to me will be a bit problematic since I am actually working on a little something to generate front end representations of java objects/classes, which are classes generated by an ORM, so each table on the DB having an "id" column produces a class with an id attribute on java, which now I am seeing reflected on my WebComponents. I guess I'll need to add a couple more if's to handle that (and each of the other attributes on the page you shared) :smiling_face_with_tear:

**Reply by ynxtynxt** [reply]
How about generating a pair of Reactive Controller/element instead so you map the properties at connecting time?

**Reply by Ordiel** [reply]
That sounds like a good idea, will give it a go!
